src/sentry/servermode.py

```diff
@@ -0,0 +1,108 @@
+from __future__ import annotations
+
+import abc
+import itertools
+from enum import Enum
+from typing import Any, Callable, Iterable
+
+from django.conf import settings
+
+
+class ServerComponentMode(Enum):
+    """Defines the component mode of the application to be acting as."""
+
+    MONOLITH = "MONOLITH"
+    CONTROL = "CONTROL"
+    CUSTOMER = "CUSTOMER"
+    FRONTEND = "FRONTEND"
+
+    @classmethod
+    def resolve(cls, mode: str | ServerComponentMode | None) -> ServerComponentMode:
+        if not mode:
+            return cls.MONOLITH
+        if isinstance(mode, ServerComponentMode):
+            return mode
+
+        try:
+            obj = getattr(cls, mode)
+            if isinstance(obj, cls):
+                return obj
+        except AttributeError:
+            pass
+
+        raise ValueError(f"Not a {cls.__name__} name: {mode!r}")
+
+    def __str__(self) -> str:
+        return self.value
+
+    @classmethod
+    def get_current_mode(cls) -> ServerComponentMode:
+        return cls.resolve(settings.SERVER_COMPONENT_MODE)
+
+
+class ModeLimited(abc.ABC):
```

Comment:
Do we plan to apply the `ModeLimited` decorator to `Service` classes as well ?
https://github.com/getsentry/sentry/blob/fbe987561d2b9df070f34652785294cc430b41e4/src/sentry/utils/services.py#L44

When we access storages that are not postgres model (like nodestore in production) we rely on that abstraction.
Or do we plan to have that covered by limiting api endpoints ?

Reply:
>  Or do we plan to have that covered by limiting api endpoints ?

We should be able to cover nodestore, evenstore, evenstream and tdsb usage with endpoint limiting.

Reply:
🤔 Sounds like we might need to make the service abstractions for storage systems aware of the deployment mode as well.